src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,485 @@
+
+-- | Overlay a given list of graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- overlays []          == 'empty'
+-- overlays [x]         == x
+-- overlays [x,y]       == 'overlay' x y
+-- overlays             == 'foldr' 'overlay' 'empty'
+-- 'isEmpty' . overlays == 'all' 'isEmpty'
+-- @
+overlays :: Ord a => [SymmetricRelation a] -> SymmetricRelation a
+overlays = SymmetricRelation . R.overlays . map fromSymmetric
+
+-- | Connect a given list of graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- connects []          == 'empty'
+-- connects [x]         == x
+-- connects [x,y]       == 'connect' x y
+-- connects             == 'foldr' 'connect' 'empty'
+-- 'isEmpty' . connects == 'all' 'isEmpty'
+-- connects             == connects . 'reverse
+-- @
+connects :: Ord a => [SymmetricRelation a] -> SymmetricRelation a
+connects = foldr connect empty
+
+-- | The 'isSubgraphOf' function takes two graphs and returns 'True' if the
+-- first graph is a /subgraph/ of the second.
+-- Complexity: /O((n + m) * log(n))/ time.
+--
+-- @
+-- isSubgraphOf 'empty'         x               ==  True
+-- isSubgraphOf ('vertex' x)    'empty'         ==  False
+-- isSubgraphOf x             ('overlay' x y)   ==  True
+-- isSubgraphOf ('overlay' x y) ('connect' x y) ==  True
+-- isSubgraphOf ('path' xs)     ('circuit' xs)  ==  True
+-- isSubgraphOf x y                             ==> x <= y
+-- isSubgraphOf (edge x y) (edge y x)           == True
+-- @
+isSubgraphOf :: Ord a => SymmetricRelation a -> SymmetricRelation a -> Bool
+isSubgraphOf x y = R.isSubgraphOf (fromSymmetric x) (fromSymmetric y)
+
+-- | Check if a relation is empty.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- isEmpty 'empty'                         == True
+-- isEmpty ('overlay' 'empty' 'empty')     == True
+-- isEmpty ('vertex' x)                    == False
+-- isEmpty ('removeVertex' x $ 'vertex' x) == True
+-- isEmpty ('removeEdge' x y $ 'edge' x y) == False
+-- @
+isEmpty :: SymmetricRelation a -> Bool
+isEmpty = R.isEmpty . fromSymmetric
+
+-- | Check if a graph contains a given vertex.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasVertex x 'empty'            == False
+-- hasVertex x ('vertex' x)       == True
+-- hasVertex 1 ('vertex' 2)       == False
+-- hasVertex x . 'removeVertex' x == 'const' False
+-- @
+hasVertex :: Ord a => a -> SymmetricRelation a -> Bool
+hasVertex x = R.hasVertex x . fromSymmetric
+
+-- | Check if a graph contains a given edge.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasEdge x y 'empty'            == False
+-- hasEdge x y ('vertex' z)       == False
+-- hasEdge x y ('edge' x y)       == True
+-- hasEdge y x ('edge' x y)       == True
+-- hasEdge x y . 'removeEdge' x y == 'const' False
+-- hasEdge x y                    == 'elem' (x,y) . 'edgeList'
```

Comment:
Is this really true? What if `x = 1` and `y = 0`? I think this property will fail in this case. Has it been tested?

Reply:
If you're talking about the first property it is being checked. The last property is also being checked on the next update and it is also true.

Reply:
I'm talking about `hasEdge x y == 'elem' (x,y) . 'edgeList'`. Try `hasEdge 1 0 $ edge 0 1`, and then try `elem (1,0) $ edgeList $ edge 0 1`.

Reply:
Indeed it's not true... strangely the tests passed last time I checked. should I remove this property?

Reply:
I suggest to switch to `elem (min x y, max x y)` which should be correct, yet still readable.

Reply:
That's clever!